src/policy/marksweepspace/native_ms/global.rs

```diff
@@ -0,0 +1,366 @@
+        self.pr.release_pages(block.start());
+    }
+
+    pub fn block_clear_metadata(&self, block: Block) {
+        for metadata_spec in Block::METADATA_SPECS {
+            metadata_spec.set_zero_atomic(block.start(), Ordering::SeqCst);
+        }
+        #[cfg(feature = "global_alloc_bit")]
+        crate::util::alloc_bit::bzero_alloc_bit(block.start(), Block::BYTES);
+    }
+
+    pub fn acquire_block(&self, tls: VMThread, size: usize, align: usize) -> BlockAcquireResult {
+        {
+            let mut abandoned = self.abandoned.lock().unwrap();
+            let bin = mi_bin::<VM>(size, align);
+
+            {
+                let abandoned_available = &mut abandoned.available;
+                if !abandoned_available[bin].is_empty() {
+                    let block = abandoned_available[bin].pop().unwrap();
+                    return BlockAcquireResult::AbandonedAvailable(block);
+                }
+            }
+
+            {
+                let abandoned_unswept = &mut abandoned.unswept;
+                if !abandoned_unswept[bin].is_empty() {
+                    let block = abandoned_unswept[bin].pop().unwrap();
+                    return BlockAcquireResult::AbandonedUnswept(block);
+                }
+            }
+        }
+
+        let acquired = self.acquire(tls, Block::BYTES >> LOG_BYTES_IN_PAGE);
+        if acquired.is_zero() {
+            BlockAcquireResult::Exhausted
+        } else {
+            BlockAcquireResult::Fresh(Block::from_unaligned_address(acquired))
+        }
+    }
+
+    pub fn generate_sweep_tasks(&self) -> Vec<Box<dyn GCWork<VM>>> {
+        // # Safety: ImmixSpace reference is always valid within this collection cycle.
+        let space = unsafe { &*(self as *const Self) };
+        self.chunk_map
+            .generate_tasks(|chunk| Box::new(SweepChunk { space, chunk }))
+    }
+}
+
+use crate::scheduler::GCWork;
+use crate::MMTK;
+
+/// Chunk sweeping work packet.
+struct SweepChunk<VM: VMBinding> {
+    space: &'static MarkSweepSpace<VM>,
+    chunk: Chunk,
+}
+
+impl<VM: VMBinding> GCWork<VM> for SweepChunk<VM> {
+    #[inline]
+    fn do_work(&mut self, _worker: &mut GCWorker<VM>, _mmtk: &'static MMTK<VM>) {
+        debug_assert!(self.space.chunk_map.get(self.chunk) == ChunkState::Allocated);
+        // number of allocated blocks.
+        let mut allocated_blocks = 0;
+        // Iterate over all allocated blocks in this chunk.
+        for block in self
+            .chunk
+            .iter_region::<Block>()
+            .filter(|block| block.get_state() != BlockState::Unallocated)
+        {
+            if !block.attempt_release(self.space) {
+                // Block is live. Increment the allocated block count.
+                allocated_blocks += 1;
+            }
+        }
+        // Set this chunk as free if there is not live blocks.
+        if allocated_blocks == 0 {
+            self.space.chunk_map.set(self.chunk, ChunkState::Free)
+        }
+    }
+}
```

Comment:
I have some radical thoughts about this.  Maybe this `SweepChunk` work packet is completely unnecessary.

If each `BlockList` is owned by exactly one mutator (or the global `MarkSweepSpace`), and each `Block` is owned by exactly one `BlockList`, then we can traverse all existing blocks simply by traversing all `BlockList` instances.  That means visiting all `BlockList` in mutators and the lists in `MarkSweepSpace::abandoned_lists`.  Doing so allows blocks to be swept while the current thread has exclusive access to its `BlockList`, therefore removing the need to have `lock()` in `BlockList` completely.

But I prefer doing this under the protection of Rust's ownership model.  From the current code, it looks like there is no way for `Block` instances to be created anywhere else (other than `MarkSweepSpace::acquire_block`), held anywhere else, or discarded before they are swept.  Rust's ownership model can enforce that, but currently our `Block` type implements `Copy`, has pointer semantics, and bypasses the ownership model.  I elaborated my thoughts here: https://github.com/mmtk/mmtk-core/issues/696.